### PR TITLE
New Menu

### LINKED ISSUE #187: Explorer menu redesign
**Description**
Add a menu similar to the Cow Swap one.
image

Options in this first iteration:

- Home
- More
    > Cow Protocol
    > Documentation
    > Analytics
    > Discord
    > App Data

**Alternatives**
Next iteration: add Solver page access.

**Additional context**
https://github.com/cowprotocol/explorer/issues/156#issuecomment-1195968654

### PR DESCRIPTION
**Pull request by alongoni**
Closes cowprotocol/explorer#187 

- [x] Add Menu component to Storybook
- [x] Review styles on Storybook
- [x] Adapt the content, constants and routes
- [x] Mobile menu
- [x] Add menu to Explorer
- [x] Add icons

# Summary
Add a menu similar to the Cow Swap one.
Options in this first iteration:
- Home
- More
    > Cow Protocol
    > Documentation
    > Analytics
    > Discord
    > AppData (Metadata)
  
Next iteration:
- Solvers

https://github.com/cowprotocol/cowswap/issues/3721
image

Preview: 
- Explorer: https://pr186--explorer.review.gnosisdev.com/
- Storybook https://pr186--explorer.review.gnosisdev.com/storybook/?path=/story/common-menu--main-menu
-

### CODE REVIEW

**Comment by elena-zh** [comment]
Hey @alongoni , I know that the PR is still WIP, but I have done an early review, and here are my notes and proposals: 
1. We tend to open external links in a new tab, and internal in the same one. So it would be nice to open App data link in the same tab
in the same page
2. Scrollbar appears when the menu is changed to a burger icon
scrollbar
3. header menu is hidden when a screen width is in ~ [740-920]px
more is hidden
4. The page looks strange when[ return back from a mobile view](https://watch.screencastify.com/v/O4Eqd9WNKJGWBlN7be07) to a desktop view, menu is opened
resize
5. In the Cow Swap app sections are not displayed in a mobile view in the More section. Besides, options are displayed in the manner 'big finger UX'. Maybe we should do the same, WDYT?
no sections

Thanks!

**Comment by alongoni** [comment]
Thanks Elena for the feedback.
Point 3, is a previous issue I've found and It's in production now. It's related to the token table. https://www.loom.com/share/0cbdc8d8b48546b4a30ceb46af6ca43c

Maybe we need to create a new issue to address it.

**Comment by elena-zh** [comment]
> Thanks Elena for the feedback. Point 3, is a previous issue I've found and It's in production now. It's related to the token table. https://www.loom.com/share/0cbdc8d8b48546b4a30ceb46af6ca43c
> 
> Maybe we need to create a new issue to address it.

@alongoni , we don't need to create a new issue as it already exists on the board: #112

File: src/utils/mediaQueries.ts [hunk 1]
```diff
@@ -1,25 +1,33 @@
 import { Command } from 'types'
+import { media } from 'theme/styles/media'
 
 export type Breakpoints = 'xl' | 'lg' | 'md' | 'sm' | 'xs'
 
+const extrapolatedScreenSize = {
```
Comment (henrypalacios) [inline src/utils/mediaQueries.ts:6]: @alongoni I have taken as `media screen sizes` for the `Breakpoints` the same ones used in the `styles`, this should solve what was driving you crazy.

**Comment by henrypalacios** [comment]
@alongoni, @elena-zh up to this point, I think the thing missing is to call the routes (on code) with the icons in the `Dropdown` menu.

File: src/apps/explorer/layout/Header.tsx [hunk 2]
```diff
@@ -25,45 +30,17 @@ export const Header: React.FC = () => {
...
-    setBarActive(false)
-    history.push(`/${prefixNetwork || ''}`)
-  }
-
   return (
-    <GenericHeader logoAlt="CoW Protocol Explorer" linkTo={`/${prefixNetwork || ''}`}>
+    <GenericHeader
+      logoAlt="CoW Protocol Explorer"
+      linkTo={`/${prefixNetwork || ''}`}
+      onClickOptional={handleMobileMenuOnClick}
```
Comment (henrypalacios) [inline src/apps/explorer/layout/Header.tsx:37]: Not sending the click function when the menu is closed would avoid opening the mobile menu when clicking on the logo.

```typescript
onClickOptional={isMobileMenuOpen ? handleMobileMenuOnClick : undefined}
```

File: src/apps/explorer/layout/Header.tsx [hunk 3]
```diff
@@ -1,22 +1,27 @@
...
-  const [isBarActive, setBarActive] = useState(false)
-  const flexWrapDivRef = createRef<HTMLDivElement>()
-  useOnClickOutside(flexWrapDivRef, () => isBarActive && setBarActive(false))
+  const isMobile = useMediaBreakpoint(['xs', 'sm'])
+  const [isMobileMenuOpen, setMobileMenuOpen] = useState(false)
+
+  // Toggle the 'noScroll' class on body, whenever the mobile menu or orders panel is open.
+  // This removes the inner scrollbar on the page body, to prevent showing double scrollbars.
+  useEffect(() => {
+    isMobileMenuOpen ? addBodyClass('noScroll') : removeBodyClass('noScroll')
```
[the remark this answers is not in the document]
Comment (henrypalacios) [reply src/apps/explorer/layout/Header.tsx:19]: @alongoni :point_up:

**Review by henrypalacios — CHANGES_REQUESTED** [review]

File: src/apps/explorer/layout/Header.tsx [hunk 4]
```diff
@@ -1,22 +1,27 @@
...
+  }, [isMobileMenuOpen, isMobile])
+
+  const handleMobileMenuOnClick = useCallback(() => {
+    isMobile && setMobileMenuOpen(!isMobileMenuOpen)
+  }, [isMobile, isMobileMenuOpen])
```
Comment (ramirotw) [inline src/apps/explorer/layout/Header.tsx:24]:
```suggestion
  const handleMobileMenuOnClick = useCallback(() => {
    isMobile && setMobileMenuOpen(isMobileMenuOpen => !isMobileMenuOpen)
  }, [isMobile])
```

File: src/components/common/MenuDropdown/InternalExternalLink.tsx [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+        {menuImage}
+        {title}
+        {menuImageExternal}
+      </ExternalLink>
+    )
+  } else {
+    return (
+      <LinkWithPrefixNetwork
+        to={url}
+        rel="noopener noreferrer"
```
Comment (ramirotw) [inline src/components/common/MenuDropdown/InternalExternalLink.tsx:47]: shouldn't the `noopener noreferrer` be set for external links?

File: src/components/common/MenuDropdown/MenuTree.tsx [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+
+import InternalExternalMenuLink from 'components/common/MenuDropdown/InternalExternalLink'
+import { MAIN_MENU } from 'components/common/MenuDropdown/mainMenu'
+import { Wrapper, MenuContainer } from 'components/common/MenuDropdown/styled'
+import { MenuItemKind, MenuTreeItem } from 'components/common/MenuDropdown/types'
+import DropDown from '.'
+
+export interface ContextProps {
+  handleMobileMenuOnClick: () => void
+  isMobileMenuOpen: boolean
```
Comment (ramirotw) [inline src/components/common/MenuDropdown/MenuTree.tsx:11]: is this necessary? isn't the same as `MenuTreeProps`?

File: src/components/common/MenuDropdown/MenuTree.tsx [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+  return (
+    <Wrapper isMobileMenuOpen={isMobileMenuOpen}>
+      <MenuContainer className={isMobileMenuOpen ? 'mobile-menu' : ''}>
+        {MAIN_MENU.map((menuItem, index) => (
+          <MenuItemWithDropDown key={index} menuItem={menuItem} context={context} />
+        ))}
+      </MenuContainer>
+    </Wrapper>
+  )
+}
```
Comment (ramirotw) [inline src/components/common/MenuDropdown/MenuTree.tsx:50]:
```suggestion
export function MenuTree(props: MenuTreeProps): JSX.Element {
  const { handleMobileMenuOnClick, isMobileMenuOpen } = props
  return (
    <Wrapper isMobileMenuOpen={isMobileMenuOpen}>
      <MenuContainer className={isMobileMenuOpen ? 'mobile-menu' : ''}>
        {MAIN_MENU.map((menuItem, index) => (
          <MenuItemWithDropDown key={index} menuItem={menuItem} context={props} />
        ))}
      </MenuContainer>
    </Wrapper>
  )
}
```

**Comment by elena-zh** [comment]
Hey @alongoni , @henrypalacios , great job!
Some my notes/edge cases: 
1. 
> > Thanks Elena for the feedback. Point 3, is a previous issue I've found and It's in production now. It's related to the token table. https://www.loom.com/share/0cbdc8d8b48546b4a30ceb46af6ca43c
> > Maybe we need to create a new issue to address it.
> 
> @alongoni , we don't need to create a new issue as it already exists on the board: #112
I believe, it is going to be fixed in another PR
2. [Menu is opened](https://watch.screencastify.com/v/oEvbBq4WTQCf0SJl7ALm) when I press on the Cow Logo. Moreover, the logo hangs in a 'moving' state in a mobile device when I press on it
image
3. More option disappears when I move to the landscape mode in a mobile device: 

https://user-images.githubusercontent.com/70885163/189333640-4d7c5934-1665-4819-a07d-77e04c9ab4f5.mp4

4. The page looks strange when[ return back from a mobile view](https://watch.screencastify.com/v/O4Eqd9WNKJGWBlN7be07) to a desktop view, menu is opened, screen width is ~ [730-1020]px. When screen width is >1020 px, there is no menu at all (I mean, when continue expanding the page)
menu is missing
5. When I open a menu in Win10 in a reduced screen, unnecessary scrollbar is displayed there
scrollbar in windows


Thanks!

File: src/components/layout/GenericLayout/Header/index.tsx [hunk 8]
```diff
@@ -67,11 +71,12 @@ const Logo = styled(Link)`
 type Props = PropsWithChildren<{
   linkTo?: string
   logoAlt?: string
+  onClickOptional?: React.MouseEventHandler<HTMLAnchorElement>
```
Comment (ramirotw) [inline src/components/layout/GenericLayout/Header/index.tsx:74]: `onClickOptional` why the Optional? isn't enough to set the prop as nullable?

**Review by ramirotw — COMMENTED** [review]
really nice!! 😍

File: src/components/common/MenuDropdown/InternalExternalLink.tsx (hunk 5 above)
Comment (alongoni) [reply src/components/common/MenuDropdown/InternalExternalLink.tsx:47]: Exactly! External links already have it. Removed for internal links.

**Comment by elena-zh** [comment]
Hey @alongoni , changes LGTM in Explorer. 

However, I have found issues in Storybook: 
 - The page is crashed when I expand the menu
crash on expand
2. Menu looks strange in a mobile view
mobile

Could you please fix it?

Thanks

**Review by matextrem — APPROVED** [review]

**Comment by elena-zh** [comment]
Hey @alongoni , now the menu is not crashed in Storybook when open it. 
Just a tiny question: is this OK that in a mobile view I do not see a burger icon that opens a menu, and the menu layout is different than in Explorer?
image

It looks to me that in a desktop view we show the whole menu now, and in a mobile view a menu component ('more'). Maybe we shouldn't mix these things?

Thanks

**Review by elena-zh — APPROVED** [review]
Hey @alongoni , great! 
I'm approving the issue in order not to block the release, however still have 1 nitpick: could we please move the burger menu to the right?
move

Thanks!

**Review by ramirotw — APPROVED** [review]
😍

File: src/apps/explorer/layout/Header.tsx (hunk 3 above)
[the remark this answers is not in the document]
Comment (alongoni) [reply src/apps/explorer/layout/Header.tsx:18]: We need this to prevent double scrolling when the menu is open on mobile. Thanks!

**Review by henrypalacios — APPROVED** [review]

**Pull request merged**